flink-core/src/main/java/org/apache/flink/core/memory/DirectMemorySegment.java

```diff
@@ -0,0 +1,560 @@
+			
+			// this must be after the get() call to ensue that the byte buffer is not
+			// modified in case the call fails
+			source.position(sourceOffset + numBytes);
+		}
+		else {
+			// neither heap buffer nor direct buffer
+			while (source.hasRemaining()) {
+				put(offset++, source.get());
+			}
+		}
+	}
+	
+	@Override
+	@SuppressWarnings("restriction")
+	public final void copyTo(int offset, MemorySegment target, int targetOffset, int numBytes) {
+		if (target.getClass() == DirectMemorySegment.class) {
+			DirectMemorySegment directOther = (DirectMemorySegment) target;
+			
+			final long thisPointer = address + offset;
+			final long otherPointer = directOther.address + targetOffset;
+			
+			if (numBytes >= 0 && thisPointer <= addressLimit - numBytes && otherPointer <= directOther.addressLimit - numBytes) {
+				UNSAFE.copyMemory(thisPointer, otherPointer, numBytes);
+			}
+			else if (address > addressLimit || directOther.address > directOther.addressLimit) {
+				throw new IllegalStateException("disposed");
+			}
+			else {
+				throw new IndexOutOfBoundsException();
+			}
+		}
+		else {
+			throw new IllegalArgumentException("Can only copy to other direct memory segments");
+		}
+	}
+
+	// -------------------------------------------------------------------------
+	//                      Comparisons & Swapping
+	// -------------------------------------------------------------------------
+
+	@Override
+	public int compare(MemorySegment seg2, int offset1, int offset2, int len) {
+		ByteBuffer b1 = this.buffer;
+		ByteBuffer b2 = ((DirectMemorySegment)seg2).buffer;
+		return b1.compareTo(b2);
+	}
+
+	@Override
+	public void swapBytes(MemorySegment seg2, int offset1, int offset2, int len) {
+		while(len >= 8) {
+			long tmp = this.getLong(offset1);
+			this.putLong(offset1, seg2.getLong(offset2));
+			seg2.putLong(offset2, tmp);
+			offset1+=8;
+			offset2+=8;
+			len-=8;
+		}
+		while(len > 0) {
+			byte tmp = this.get(offset1);
+			this.put(offset1, seg2.get(offset2));
+			seg2.put(offset2, tmp);
+			offset1++;
+			offset2++;
+			len--;
+		}
+
+	}
+
+	// --------------------------------------------------------------------------------------------
+	//                     Utilities for native memory accesses and checks
+	// --------------------------------------------------------------------------------------------
+	
+	@SuppressWarnings("restriction")
+	private static final sun.misc.Unsafe UNSAFE = MemoryUtils.UNSAFE;
+	
+	@SuppressWarnings("restriction")
+	private static final long BYTE_ARRAY_BASE_OFFSET = UNSAFE.arrayBaseOffset(byte[].class);
+	
+	private static final long COPY_PER_BATCH = 1024 * 1024;
+	
+	private static final Field ADDRESS_FIELD;
+	
+	static {
+		try {
+			ADDRESS_FIELD = java.nio.Buffer.class.getDeclaredField("address");
+			ADDRESS_FIELD.setAccessible(true);
+		}
+		catch (Throwable t) {
+			throw new RuntimeException("Cannot initialize DirectMemorySegment - direct memory not supported by Flink");
```

Comment:
I would suggest to add a hint to the error message suggesting to turn off the direct memory allocation.

(I actually find the last part of the message confusing, because it is actually not really "not supported by Flink", but by the system?)